cluster-autoscaler/proposals/provisioning-request.md

```diff
@@ -0,0 +1,293 @@
+            pods/nodes.
+          - Size limiting, which caps the scale-up to match the size
+            restrictions (on node group or cluster level).
+3.  Node creation, which needs to support atomic resize. Either via native cloud
+    provider APIs or best effort with node removal if CA is unable to fulfill
+    the scale-up.
+      - This is also quite substantial change, we can provide a generic
+        best-effort implementation that will try to scale up and clean-up nodes
+        if it is unsuccessful, but it is up to cloud providers to integrate with
+        provider specific APIs.
+4.  Scale down path is not expected to change much. But users should follow
+    [best
+    practices](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node)
+    to avoid CA disturbing their workloads.
+
+## Testing
+
+The following e2e test scenarios will be created to check whether ProvReq
+handling works as expected:
+
+1.  A new ProvReq with `GenericCheckCapacity` provisioning class is created, CA
+    checks if there is enough capacity in cluster to provision specified pods.
+2.  A new ProvReq with `GenericAtomicScaleUp` provisioning class is created, CA
+    picks an appropriate node group scales it up atomically.
+3.  A new atomic ProvReq is created for which a NAP needs to provision a new
+    node group. NAP creates it CA scales it atomically.
+      - Here we should cover some of the different reasons why NAP may be
+        required.
+4.  An atomic ProvReq fails due to node group size limits and NAP CPU and/or RAM
+    limits.
+5.  Scalability tests.
+      - Scenario in which many small ProvReqs are created (strain on the number
+        of scale-up loops).
+      - Scenario in which big ProvReq is created (strain on a single scale-up
+        loop).
+
+## Future Expansions
+
+### ProvisioningClass CRD
+
+One of the expansion of this approach is to introduce the ProvisioningClass CRD,
+which follows the same approach as
+[StorageClass object](https://kubernetes.io/docs/concepts/storage/storage-classes/).
+Such approach would allow administrators of the cluster to introduce a list of allowed
+ProvisioningClasses. Such CRD can also contain a pre set configuration, i.e.
+administrators may set that `GenericAtomicScaleUp` would retry up to `2h`.
+
+Possible CRD definition:
+```go
+// ProvisioningClass is a way to express provisioning classes available in the cluster.
+type ProvisioningClass struct {
+	// Name denotes the name of the object, which is to be used in the ProvisioningClass 
+	// field in Provisioning Request CRD.
+	// 
+	// +kubebuilder:validation:Required
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	// AdditionalParameters contains all other parameters custom classes may require.
+	//
+	// +optional
+	AdditionalParameters map[string]string `json:"additionalParameters" protobuf:"bytes,2,name=additionalParameters"`
```

Comment:
how do these differ from the AdditionalParameters in the ProvisioningRequest?

Reply:
Those would allow cluster admins to specify cluster wide defaults (that users either could override or would be enforced).